[P099] Touch XPT2046 new plugin

tonhuisman opened the pull request:
- Adds Touch support for XPT2046 (often combined with ILI9341 TFT)
- Space for up to 40 user defined touchobjects by name
- Allows several types of events, X/Y, X/Y/Z, objectname [TaskName#ObjectName]
- Calibration to map to screen coordinates (optional)

TODO:
- [x] ~~Implement VSPI support on ESP32~~ (connected to VSPI pins and all works as expected)
- [x] Documentation
- [x] Add to some pre-configured build setup? (`#define USES_P099`)
- [x] Add touch-button feature
- [x] Add commands for enabling/disabling touch objects

NB: Needs PR #3205

tonhuisman committed:
[P099] Touch XPT2046 new plugin
- Adds Touch support for XPT2046 (often combined with ILI9341 TFT)
- Space for up to 40 user defined touchobjects by name
- Allows several types of events, X/Y, X/Y/X, objectname [TaskName#ObjectName]
- Calibration to map to screen coordinates (optional)

TODO:
- Implement VSPI support on ESP32
- Documentation

NB: Needs PR #3205

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

TD-er commented on src/_P099_XPT2046Touch.ino:
```diff
@@ -0,0 +1,450 @@
+        addFormNumericBox(F("Screen Height (px) (y)"), F("p099_height"), height_, 1, 65535);
+
+        byte choice2 = P099_CONFIG_ROTATION;
+        String options2[4] = { F("Normal"), F("+90°"), F("+180°"), F("+270°") };
```
Please don't use unicode in the source code as encoded character in the string.
If we need to have a degree symbol, then we should add a function for it.
That may also provide the right encoding either for html (default) or if needed at all one encoded to be sent via serial or something.

tonhuisman replied:
I 'borrowed' that from P095...

TD-er replied:
Then that one also needs some de-Unicoding :)

TD-er commented on src/_P099_XPT2046Touch.ino:
```diff
@@ -0,0 +1,450 @@
+        addFormSelector(F("Events"), F("p099_events"), 6, options3, optionValues3, choice3);
+
+        {
+          P099_data_struct *P099_data = new (std::nothrow) P099_data_struct(); //static_cast<P099_data_struct *>(getPluginTaskData(event->TaskIndex));
```
Just curious, why do you create a new data struct in the `PLUGIN_WEBFORM_LOAD` call instead of the `PLUGIN_INIT` call?
If the plugin is enabled if you load the page, then this object will be present in memory twice.

TD-er commented on src/_P099_XPT2046Touch.ino:
```diff
@@ -0,0 +1,450 @@
+            addFormSelector(F("# of objects"), F("p099_objectcount"), 6, options5, optionValues5, choice5, true);
+          }
+          if (P099_CONFIG_OBJECTCOUNT > -1) {
+            {
```
Why the extra scope (the extra level of curly braces) ?
There is also an extra level at the for loop in line 216.

TD-er commented on src/src/PluginStructs/P099_data_struct.h:
```diff
@@ -0,0 +1,98 @@
+  bool isValidAndTouchedTouchObject(uint16_t x, uint16_t y, String &selectedObjectName, uint8_t checkObjectCount);
+  void scaleRawToCalibrated(uint16_t &x, uint16_t &y);
+
+  XPT2046_Touchscreen *touchscreen;
```
You also store the pointer.
Just make sure this pointer is set to `nullptr` after it is being read or else bad things may happen.

TD-er commented on src/src/PluginStructs/P099_data_struct.h:
```diff
@@ -0,0 +1,98 @@
+typedef struct {
+  tP099_Point top_left;
+  tP099_Point bottom_right;
+} tP099_Calibration;
```
These are not members of the data_struct object, which means they are always present in the global scope.
Regardless whether the plugin is active or not and thus will also be shared among multiple instances of this plugin.

If you need to have a struct for saving, then make a separate struct for it and declare that as a member of the data_struct object.
When declared like this `typedef struct { ... } var_name;` you define and declare the variable and since it is outside the scope of a struct or class, it is a variable in the global scope.

TD-er reviewed (commented):
Just a few comments of the first few things that caught my eye.
I think it is best to split the struct to store into a separate struct and declare it as a variable in the data_struct for this plugin.
Also include all global data you need into this data_struct as it is active data for that instance of the plugin.
That's the whole reason I created this `PluginTaskData_base` inheritance structure, so there is no active data in memory when a plugin is not enabled and also it separates data per instance of a plugin (thus a task), allowing multiple instances of the same plugin.

tonhuisman committed:
[P099] Code improvements
- Avoid using unicode in source
- Less scope braces
- Memory optimizations
- Improve data struct definition
- Pointer initializations
- Cleanup on `PLUGIN_EXIT`

tonhuisman commented:
I resolved all remarks, and if the unicode resolution is as expected, can also fix the same issue for P095 and P096.

Now lets fix the TODO's 😄

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Cut off x/y coordinate outside calibration

tonhuisman committed:
[P099] Documentation for TouchScreen
- Device configuration
- Setup calibration
- Define TouchObjects

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[Docs] Add Touch plugin category
- [P099] Improvements and corrections
- [P095] Adjust documententation title (remove XPT2046)

tonhuisman committed:
Merge branch 'feature/XPT2046_touchscreen' of https://github.com/tonhuisman/ESPEasy-1 into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Events only fired when Rules enabled
Show note in Device settings page if Rules disabled
Add USES_P099 to `define_plugin_sets.h` (disabled)

tonhuisman commented:
@TD-er This PR seems to be complete now, can you review please, so I can resolve any remarks if needed?

Others are of course also welcome to review 😄

tonhuisman committed:
[P099] Fixed an issue in x/y scaling function

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman commented:
> @TD-er This PR seems to be complete now

I am working on some more changes, also to the data structure, so this needs some more baking before it's being merged.

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'feature/XPT2046_touchscreen' of https://github.com/tonhuisman/ESPEasy-1 into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Add On/Off button to touchobjects
- Add On/Off button and Inverted settings per touchobject
- Updated documentation and screenshots

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Settings page to show when task is disabled

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

TD-er committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Adjust #include to new structure, enable plugin in test build

tonhuisman commented:
Enabled plugin in `define_plugin_sets.h` so Travis can actually try to compile it 😎 , though the display plugin it is often accompanied by (P095) isn't included. Had to turn off another plugin to make this one fit in the `test` build-set (ESP8266) 😞

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Improve string allocation, add options, source formatting
- Source formatting and String allocation using reserve() for longer strings
- Add option and command for flipping the rotation setting
- Improved rotation setting to conform to ILI9341 setting
- Add option for setting the debounce delay for On/Off buttons

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
[P099] Add updated file

TD-er commented on src/src/PluginStructs/P099_data_struct.cpp:
```diff
@@ -262,7 +279,7 @@ void P099_data_struct::scaleRawToCalibrated(uint16_t &x, uint16_t &y) {
         _y = StoredSettings.Calibration.bottom_right.y;
       }
-      float y_fact = (StoredSettings.Calibration.bottom_right.y - StoredSettings.Calibration.top_left.y) / ts_y_res;
+      float y_fact = (StoredSettings.Calibration.bottom_right.y - StoredSettings.Calibration.top_left.y) / _ts_y_res;
```
Why store it in a float here?
All values are int types and not casted to a float, so it will only be converted to a float when assigning it, not during computation.

tonhuisman replied:
Think I'm spoiled by all the Java programming I did in the past 🤔 will investigate and fix.

TD-er commented on src/src/PluginStructs/P099_data_struct.cpp:
```diff
@@ -252,7 +269,7 @@ void P099_data_struct::scaleRawToCalibrated(uint16_t &x, uint16_t &y) {
+      float x_fact = (StoredSettings.Calibration.bottom_right.x - StoredSettings.Calibration.top_left.x) / _ts_x_res;
       x = int((_x * 1.0f) / x_fact);
     }
     uint16_t _y = y - StoredSettings.Calibration.top_left.y;
```
`_y` is not a member here, but just a local variable.
So why using an underscore as prefix here?   (same for `_x`)

tonhuisman replied:
This was written before we chose the underscore as prefix for member variables. Will fix.

TD-er commented on src/src/PluginStructs/P099_data_struct.cpp:
```diff
@@ -252,7 +269,7 @@ void P099_data_struct::scaleRawToCalibrated(uint16_t &x, uint16_t &y) {
       }
-      float x_fact = (StoredSettings.Calibration.bottom_right.x - StoredSettings.Calibration.top_left.x) / ts_x_res;
+      float x_fact = (StoredSettings.Calibration.bottom_right.x - StoredSettings.Calibration.top_left.x) / _ts_x_res;
       x = int((_x * 1.0f) / x_fact);
```
Multiplying by `1.0f` is not really needed here as you divide it by a float (which may not be needed, see other comment on `y_fact`)
If you need to cast it to a float, use `static_cast<float>(_x)`
Also using it as a constructor for `int` isn't really readable. Better use `static_cast<int>(...)`.

N.B. casting it to an int like this is effectively the same as `floor`. Is that what you intended?

tonhuisman committed:
[P099] Fix casting and rounding in calculations

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'feature/XPT2046_touchscreen' of https://github.com/tonhuisman/ESPEasy-1 into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

TD-er committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

TD-er commented:
What is the status of this PR?

tonhuisman commented:
> What is the status of this PR?

Ready to merge, IMHO

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen

tonhuisman commented:
I'll asap fix the #include that is failing in the last builds (requirement for build adjustment copying cpp files into 1).

tonhuisman committed:
Merge branch 'feature/XPT2046_touchscreen' of https://github.com/tonhuisman/ESPEasy-1 into feature/XPT2046_touchscreen

tonhuisman committed:
Adjust #include path to fix build

tonhuisman committed:
Merge branch 'mega' into feature/XPT2046_touchscreen
